cmd/anonymizer/main.go

```diff
@@ -15,13 +15,106 @@
 package main
 
 import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/spf13/cobra"
 	"go.uber.org/zap"
+	"google.golang.org/grpc"
 
-	"github.com/jaegertracing/jaeger/cmd/anonymizer/app/writer"
+	app "github.com/jaegertracing/jaeger/cmd/anonymizer/app"
+	writer "github.com/jaegertracing/jaeger/cmd/anonymizer/app/writer"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/pkg/version"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
 )
 
+var logger, _ = zap.NewDevelopment()
+
+type grpcClient struct {
+	api_v2.QueryServiceClient
+	conn *grpc.ClientConn
+}
+
+func newGRPCClient(addr string) *grpcClient {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure())
+	if err != nil {
+		logger.Fatal("failed to connect with the jaeger-query service", zap.Error(err))
+	}
+
+	return &grpcClient{
+		QueryServiceClient: api_v2.NewQueryServiceClient(conn),
+		conn:               conn,
+	}
+}
+
 func main() {
-	// TODO
-	_, _ = writer.New(writer.Config{}, zap.NewNop())
-	println("not implemented")
+	var command = &cobra.Command{
+		Use:   "jaeger-anonymizer",
+		Short: "Jaeger anonymizer hashes fields of a trace for easy sharing",
+		Long:  `Jaeger anonymizer queries Jaeger query for a trace, anonymizes fields, and store in file`,
+		Run: func(cmd *cobra.Command, args []string) {
+			prefix := app.OutputDir + "/" + app.TraceID
+			conf := writer.Config{
+				MaxSpansCount:  app.MaxSpansCount,
+				CapturedFile:   prefix + ".orig",
+				AnonymizedFile: prefix + ".anon",
+				MappingFile:    prefix + ".map",
+			}
+
+			writer, err := writer.New(conf, logger, app.HashStandardTags, app.HashCustomTags, app.HashLogs, app.HashProcess)
+			if err != nil {
+				logger.Error("error while creating writer object", zap.Error(err))
+			}
+
+			queryEndpoint := app.QueryGRPCHost + ":" + strconv.Itoa(app.QueryGRPCPort)
+			logger.Info(queryEndpoint)
+
+			client := newGRPCClient(queryEndpoint)
+			defer client.conn.Close()
+
+			traceID, err := model.TraceIDFromString(app.TraceID)
+			if err != nil {
+				logger.Fatal("failed to convert the provided trace id", zap.Error(err))
+			}
+			logger.Info(app.TraceID)
+
+			response, err := client.GetTrace(context.Background(), &api_v2.GetTraceRequest{
+				TraceID: traceID,
+			})
+			if err != nil {
+				logger.Fatal("failed to fetch the provided trace id", zap.Error(err))
+			}
+
+			spanResponseChunk, err := response.Recv()
+			if err == spanstore.ErrTraceNotFound {
+				logger.Fatal("failed to find the provided trace id", zap.Error(err))
+			}
+			if err != nil {
+				logger.Fatal("failed to fetch spans of provided trace id", zap.Error(err))
+			}
+
+			spans := spanResponseChunk.GetSpans()
```

Comment:
I recommend moving all gRPC-related logic into its own package `cmd/anonymizer/app/query`. In fact, in the future we could move it into top-level pkg/query and have it implement `spanstorage.Reader`